In [2]:
%matplotlib inline
import pandas as pd
import os
from sklearn.feature_extraction.text import CountVectorizer
from matplotlib import pyplot as plt
from matplotlib_venn import venn2
import tqdm
import re
import zipfile
import json

In [3]:
#COPYING NIEK VELDHUIS
file = "alltexts.csv"
etcsl = pd.read_csv(file, keep_default_na=False)
etcsl = etcsl.loc[etcsl["lang"].str.contains("sux")]  # throw out non-Sumerian words

In [4]:
#COPYING NIEK VELDHUIS
etcsl["lemma"] = etcsl.apply(lambda r: (r["cf"] + '[' + r["gw"] + ']' + r["pos"]) 
                            if r["cf"] != '' else r['form'] + '[NA]NA', axis=1)
etcsl['lemma'] = [lemma if not lemma == '[NA]NA' else '' for lemma in etcsl['lemma'] ] 
# kick out empty forms
etcsl["lemma"] = etcsl["lemma"].str.lower()

In [84]:
#Create ETCSL catalogue
etcsl_cat = pd.DataFrame(etcsl.groupby(['id_text','text_name']).agg('count')).reset_index(level=[0,1])
#etcsl_cat = etcsl_cat[['id_text','text_name']]
etcsl_cat = etcsl_cat.set_index('id_text')
etcsl_cat = pd.DataFrame(etcsl_cat['text_name'])
etcsl_cat

text_name
id_text                                           
c.0.1.1           Ur III catalogue from Nibru (N1)
c.0.1.2              Ur III catalogue at Yale (Y1)
c.0.2.01              OB catalogue from Nibru (N2)
c.0.2.02            OB catalogue in the Louvre (L)
c.0.2.03               OB catalogue from Urim (U1)
c.0.2.04               OB catalogue from Urim (U2)
c.0.2.05               OB catalogue from Urim (U3)
c.0.2.06              OB catalogue from Nibru (N3)
c.0.2.07    OB catalogue possibly from Zimbir (B1)
c.0.2.08              OB catalogue from Nibru (N4)
c.0.2.11   OB catalogue at Andrews University (B4)
c.0.2.12                 OB catalogue at Yale (Y2)
c.0.2.13              OB catalogue from Nibru (N6)
c.1.1.1                        Enki and Ninhursaŋa
c.1.1.2                            Enki and Ninmah
c.1.1.3                   Enki and the world order
c.1.1.4                    Enki's journey to Nibru
c.1.2.1                           Enlil and Ninlil
c.1.2.2                              Enlil and Sud
c.1.3.1                             Inana and Enki
c.1.3.2                             Inana and Ebih
c.1.3.3                     Inana and Šu-kale-tuda
c.1.3.4                            Inana and Gudam
c.1.3.5                               Inana and An
c.1.4.1        Inana's descent to the nether world
c.1.4.1.1                   Dumuzid and Ŋeštin-ana
c.1.4.1.3                  Dumuzid and his sisters
c.1.4.3                            Dumuzid's dream
c.1.4.4        Inana and Bilulu: an ulila to Inana
c.1.5.1              Nanna-Suen's journey to Nibru
...                                            ...
c.6.1.02                Proverbs: collection 2 + 6
c.6.1.03                    Proverbs: collection 3
c.6.1.04                    Proverbs: collection 4
c.6.1.05                    Proverbs: collection 5
c.6.1.07                    Proverbs: collection 7
c.6.1.08                    Proverbs: collection 8
c.6.1.09                    Proverbs: collection 9
c.6.1.10                   Proverbs: collection 10
c.6.1.11                   Proverbs: collection 11
c.6.1.12                   Proverbs: collection 12
c.6.1.13                   Proverbs: collection 13
c.6.1.14                   Proverbs: collection 14
c.6.1.15                   Proverbs: collection 15
c.6.1.16                   Proverbs: collection 16
c.6.1.17                   Proverbs: collection 17
c.6.1.18                   Proverbs: collection 18
c.6.1.19                   Proverbs: collection 19
c.6.1.21                   Proverbs: collection 21
c.6.1.22                   Proverbs: collection 22
c.6.1.23                   Proverbs: collection 23
c.6.1.24                   Proverbs: collection 24
c.6.1.25                   Proverbs: collection 25
c.6.1.26                   Proverbs: collection 26
c.6.1.27                   Proverbs: collection 27
c.6.1.28                   Proverbs: collection 28
c.6.2.1                       Proverbs: from Nibru
c.6.2.2                        Proverbs: from Susa
c.6.2.3                        Proverbs: from Urim
c.6.2.4                        Proverbs: from Unug
c.6.2.5            Proverbs: of unknown provenance

[394 rows x 1 columns]

In [5]:
#COPYING NIEK VELDHUIS
file = "parsed.csv"
lexical = pd.read_csv(file, keep_default_na=False)
lexical = lexical.loc[lexical["lang"].str.contains("sux")]

In [6]:
#COPYING NIEK VELDHUIS
lexical["lemma"] = lexical.apply(lambda r: (r["cf"] + '[' + r["gw"] + ']' + r["pos"]) 
                            if r["cf"] != '' else r['form'] + '[NA]NA', axis=1)
lexical['lemma'] = [lemma if not lemma == '[NA]NA' else '' for lemma in lexical['lemma'] ] 
# kick out empty forms
lexical["lemma"] = lexical["lemma"].str.lower()

In [7]:
#COPYING NIEK VELDHUIS
lexical = lexical[~lexical["field"].isin(["sg", "pr"])] # remove lemmas that derive from the fields "sign" 
# or "pronunciation" in sign lists.

In [12]:
file = "dcclt.zip"
z = zipfile.ZipFile(file) 
st = z.read("dcclt/catalogue.json").decode("utf-8")
j = json.loads(st)
cat_df = pd.DataFrame(j["members"]).T
cat_df = cat_df[["id_text", "period","designation"]] #added designation

In [16]:
ob = cat_df[cat_df["period"] == "Old Babylonian"]
#ob[:10]

In [17]:
keep = ob.index.values
keep = ['dcclt/' + id_text for id_text in keep]
lexical = lexical.loc[lexical["id_text"].isin(keep)]

In [38]:
lexical['cat_num'] = lexical['id_text'].apply(lambda x: x.replace('dcclt/',''))

JASON STUFF

In [18]:
etcsl = etcsl[~etcsl['lemma'].str.contains('\[na\]na')]

In [19]:
lexical = lexical[~lexical['lemma'].str.contains('\[na\]na')]

In [20]:
lexical.shape

(39453, 23)

In [52]:
def compare_single(lex_wl,lit_wl):
    inter_wl = lex_wl.intersection(lit_wl)
    lex_comp = len(inter_wl) / len(lex_wl)
    lit_comp = len(inter_wl) / len(lit_wl)
    return (lex_comp,lit_comp,len(inter_wl))

In [22]:
lex_wl = set(lexical['lemma'])
lit_wl = set(etcsl['lemma'])
len(lex_wl),len(lit_wl)

(4051, 4345)

In [53]:
compare_single(lex_wl,lit_wl)

(0.7170658682634731, 0.11024165707710011, 479)

### Get Text Lists

In [39]:
lex_tl = set(lexical['cat_num'])
lex_tl = {text for text in lex_tl if 'Q' in text}
lit_tl = set(etcsl['id_text'])

In [40]:
len(lex_tl),len(lit_tl)

(14, 394)

In [56]:
compare_single(lex_wl,lit_wl)

(0.7170658682634731, 0.11024165707710011, 479)

### Compare Single Lexical Lists to entire Literary Corpus

In [89]:
lit_comps = []
lit_wl = set(etcsl['lemma'])
lit_uniq_words = len(lit_wl)
for lex_t in lex_tl:
    lex_wl = set(lexical[lexical['cat_num'] == lex_t]['lemma'])
    #print(len(lex_wl))
    lex_comp,lit_comp,match_num = compare_single(lex_wl,lit_wl)
    lit_comps.append({'lex_t':lex_t,'lex_name': cat_df.loc[lex_t,'designation'],'lit_comp':lit_comp,'lex_comp':lex_comp,
                     'lex_uniq_words':len(lex_wl),'lit_uniq_words':lit_uniq_words,'matches':match_num})
df_lit_comps = pd.DataFrame(lit_comps)

In [90]:
df_lit_comps.sort_values(by='lit_comp',ascending=False)

lex_comp                   lex_name    lex_t  lex_uniq_words  lit_comp  \
7   0.777143               OB Nippur Ea  Q000055             700  0.125201   
13  0.717066              OB Nippur Izi  Q000050             668  0.110242   
2   0.686064               OB Nippur Lu  Q000047             653  0.103107   
0   0.653358           OB Nippur Ura 01  Q000039             551  0.082854   
12  0.654886           OB Nippur Ura 02  Q000040             481  0.072497   
5   0.757895            OB Nippur Kagal  Q000048             380  0.066283   
4   0.530242           OB Nippur Ura 04  Q000041             496  0.060529   
3   0.683333               Nippur Nigga  Q000052             360  0.056617   
8   0.700000           OB Nippur Ura 03  Q000001             350  0.056387   
6   0.621333             OB Nippur Diri  Q000057             375  0.053625   
9   0.704615          OB Lu₂-azlag₂ B-C  Q000302             325  0.052704   
1   0.502747           OB Nippur Ura 05  Q000042             364  0.042117   
10  0.476190           Weidner God List  Q000263             168  0.018412   
11  0.916667  OB Nippur Ki-ulutin-bi-še  Q000045              84  0.017722   

    lit_uniq_words  matches  
7             4345      544  
13            4345      479  
2             4345      448  
0             4345      360  
12            4345      315  
5             4345      288  
4             4345      263  
3             4345      246  
8             4345      245  
6             4345      233  
9             4345      229  
1             4345      183  
10            4345       80  
11            4345       77

In [91]:
df_lit_comps.sort_values(by='lex_comp',ascending=False)

lex_comp                   lex_name    lex_t  lex_uniq_words  lit_comp  \
11  0.916667  OB Nippur Ki-ulutin-bi-še  Q000045              84  0.017722   
7   0.777143               OB Nippur Ea  Q000055             700  0.125201   
5   0.757895            OB Nippur Kagal  Q000048             380  0.066283   
13  0.717066              OB Nippur Izi  Q000050             668  0.110242   
9   0.704615          OB Lu₂-azlag₂ B-C  Q000302             325  0.052704   
8   0.700000           OB Nippur Ura 03  Q000001             350  0.056387   
2   0.686064               OB Nippur Lu  Q000047             653  0.103107   
3   0.683333               Nippur Nigga  Q000052             360  0.056617   
12  0.654886           OB Nippur Ura 02  Q000040             481  0.072497   
0   0.653358           OB Nippur Ura 01  Q000039             551  0.082854   
6   0.621333             OB Nippur Diri  Q000057             375  0.053625   
4   0.530242           OB Nippur Ura 04  Q000041             496  0.060529   
1   0.502747           OB Nippur Ura 05  Q000042             364  0.042117   
10  0.476190           Weidner God List  Q000263             168  0.018412   

    lit_uniq_words  matches  
11            4345       77  
7             4345      544  
5             4345      288  
13            4345      479  
9             4345      229  
8             4345      245  
2             4345      448  
3             4345      246  
12            4345      315  
0             4345      360  
6             4345      233  
4             4345      263  
1             4345      183  
10            4345       80

#### Observations
As we might expect, the lexical lists with more unique words overall have higher matching percentages with the literary corpus as a whole.

This is true up until OB Nippur Ura 4, which has more unique words than OB Nippur Ura 2, but less matching. This is also true about OB Nippur Diri and OB Nippur Ura 5, but to a lesser degree

---
Most of the words in Ki-ulutin-bi-še appear in some form or other in literary texts. Given the shortness and the many common words in this list, this fact is not surprising

Despite the many words contained in OB Nippur Ea and OB Nippur Izi, more of their words appear in literary texts compared to other lexical lists. This is also true of OB Nippur Kagal, but the list also has fewer unique words

The Weidner God List has the least amount of hits in the lexical corpus. More than half of the gods in this list are never attested there.

### Compare Single Literary Text to Entire Lexical Corpus

In [94]:
lex_comps = []
lex_wl = set(lexical['lemma'])
for lit_t in lit_tl:
    lit_wl = set(etcsl[etcsl['id_text'] == lit_t]['lemma'])
    #print(len(lex_wl))
    lex_comp,lit_comp,match_num = compare_single(lex_wl,lit_wl)
    lex_comps.append({'lit_t':lit_t,'lit_name':etcsl_cat.loc[lit_t]['text_name'],'lex_comp':lex_comp,'lit_comp':lit_comp,
                     'lit_uniq_words':len(lit_wl),'lex_uniq_words':len(lex_wl),'matches':match_num})
df_lex_comps = pd.DataFrame(lex_comps)

In [95]:
df_lex_comps.sort_values(by='lex_comp',ascending=False)

lex_comp  lex_uniq_words  lit_comp  \
283  0.150333            4051  0.744499   
51   0.148112            4051  0.803213   
319  0.126389            4051  0.769925   
118  0.123673            4051  0.846284   
129  0.120464            4051  0.745038   
354  0.117255            4051  0.862069   
345  0.109356            4051  0.884232   
54   0.106147            4051  0.860000   
257  0.105653            4051  0.812144   
222  0.104912            4051  0.867347   
108  0.103925            4051  0.848790   
130  0.100469            4051  0.858650   
41   0.098988            4051  0.847780   
138  0.096766            4051  0.859649   
11   0.093557            4051  0.832967   
327  0.093310            4051  0.777778   
94   0.093310            4051  0.849438   
386  0.093063            4051  0.862700   
238  0.089608            4051  0.864286   
104  0.086398            4051  0.812065   
180  0.085164            4051  0.860349   
121  0.082696            4051  0.867876   
246  0.082449            4051  0.883598   
306  0.078746            4051  0.850667   
301  0.078252            4051  0.900568   
122  0.078005            4051  0.882682   
190  0.078005            4051  0.861035   
263  0.077759            4051  0.824607   
390  0.075784            4051  0.857542   
18   0.075537            4051  0.864407   
..        ...             ...       ...   
307  0.006665            4051  0.870968   
359  0.006665            4051  0.900000   
355  0.006418            4051  0.702703   
28   0.006418            4051  0.838710   
195  0.005924            4051  0.727273   
174  0.005924            4051  0.857143   
291  0.005924            4051  0.923077   
362  0.005924            4051  0.960000   
164  0.005924            4051  0.888889   
123  0.005924            4051  1.000000   
72   0.005678            4051  0.884615   
67   0.005431            4051  0.916667   
124  0.005431            4051  0.880000   
139  0.005184            4051  0.913043   
331  0.004937            4051  0.869565   
20   0.004937            4051  0.645161   
59   0.004690            4051  0.791667   
98   0.004443            4051  1.000000   
6    0.003950            4051  0.761905   
143  0.003456            4051  0.875000   
105  0.003209            4051  1.000000   
299  0.003209            4051  0.928571   
358  0.002715            4051  0.733333   
70   0.002469            4051  0.909091   
3    0.002222            4051  0.450000   
251  0.001728            4051  0.777778   
37   0.001234            4051  0.833333   
228  0.001234            4051  1.000000   
193  0.000987            4051  0.800000   
8    0.000741            4051  0.750000   

                                              lit_name       lit_t  \
283  text_name    The building of Ninŋirsu's temple...     c.2.1.7   
51   text_name    Ninurta's exploits: a šir-sud (?)...     c.1.6.2   
319  text_name    The lament for Sumer and Urim
Nam...     c.2.2.3   
118  text_name    Lugalbanda in the mountain cave
N...   c.1.8.2.1   
129  text_name    The temple hymns
Name: c.4.80.1, ...    c.4.80.1   
354  text_name    Enki and the world order
Name: c....     c.1.1.3   
345  text_name    A praise poem of Išme-Dagan (Išme...  c.2.5.4.01   
54   text_name    The debate between Winter and Sum...     c.5.3.3   
257  text_name    Enmerkar and the lord of Aratta
N...   c.1.8.2.3   
222  text_name    Proverbs: from Urim
Name: c.6.2.3...     c.6.2.3   
108  text_name    A praise poem of Šulgi (Šulgi B)
...  c.2.4.2.02   
130  text_name    The cursing of Agade
Name: c.2.1....     c.2.1.5   
41   text_name    Proverbs: collection 3
Name: c.6....    c.6.1.03   
138  text_name    The death of Ur-Namma (Ur-Namma A...   c.2.4.1.1   
11   text_name    A hymn to Inana (Inana C)
Name: c...    c.4.07.3   
327  text_name    The lament for Urim
Name: c.2.2.2...     c.2.2.2   
94   text_name    Lugalbanda and the Anzud bird
Nam...   c.1.8.2.2   
386  text_name    Proverbs: collection 2 + 6
Name: ...    c.6.1.02   
238  tex

In [115]:
#LIMIT TO > 50 UNIQUE WORDS df_lex_comps[df_lex_comps['lit_uniq_words'] > 50]
#FEWER THAN %90 MATCH RATING
df_lex_comps[(df_lex_comps['lit_uniq_words'] > 50) & (df_lex_comps['lit_comp'] < .8)].sort_values(by='lit_comp',ascending=False)

lex_comp  lex_uniq_words  lit_comp  \
95   0.018761            4051  0.783505   
78   0.011602            4051  0.783333   
27   0.013330            4051  0.782609   
370  0.019008            4051  0.777778   
327  0.093310            4051  0.777778   
250  0.031350            4051  0.774390   
133  0.010862            4051  0.771930   
319  0.126389            4051  0.769925   
261  0.014317            4051  0.763158   
214  0.011355            4051  0.754098   
335  0.023945            4051  0.751938   
23   0.009380            4051  0.745098   
393  0.009380            4051  0.745098   
129  0.120464            4051  0.745038   
283  0.150333            4051  0.744499   
106  0.020489            4051  0.715517   
389  0.018514            4051  0.707547   
320  0.011849            4051  0.705882   
255  0.009627            4051  0.696429   
81   0.009627            4051  0.650000   
313  0.032091            4051  0.625000   
392  0.007159            4051  0.517857   
66   0.020736            4051  0.224599   

                                              lit_name      lit_t  \
95   text_name    Letter from Aradŋu to Šulgi about...   c.3.1.01   
78   text_name    A balbale to Bau for Šu-Suen (Šu-...  c.2.4.4.1   
27   text_name    A šir-namšub to Ninisina (Ninisin...   c.4.22.2   
370  text_name    A hymn to Nanna (Nanna O)
Name: c...  c.4.13.15   
327  text_name    The lament for Urim
Name: c.2.2.2...    c.2.2.2   
250  text_name    Enlil and Ninlil
Name: c.1.2.1, d...    c.1.2.1   
133  text_name    A balbale to Inana (Dumuzid-Inana...  c.4.08.01   
319  text_name    The lament for Sumer and Urim
Nam...    c.2.2.3   
261  text_name    A balbale to Nanna (Nanna B)
Name...  c.4.13.02   
214  text_name    A hymn to the E-kur
Name: c.4.80....   c.4.80.4   
335  text_name    A šir-namšub to Utu (Utu E)
Name:...   c.4.32.e   
23   text_name    A balbale to Inana for Šu-Suen (Š...  c.2.4.4.2   
393  text_name    Letter from a governor and temple...   c.3.3.05   
129  text_name    The temple hymns
Name: c.4.80.1, ...   c.4.80.1   
283  text_name    The building of Ninŋirsu's temple...    c.2.1.7   
106  text_name    Letter from Puzur-Šulgi to Šulgi ...   c.3.1.07   
389  text_name    Letter from Puzur-Šulgi to Ibbi-S...   c.3.1.19   
320  text_name    A balbale (?) to  Inana (Inana F)...   c.4.07.6   
255  text_name    A šir-namšub to Nisaba (Nisaba B)...   c.4.16.2   
81   text_name    Letter from Šamaš-ṭāb to Ilak-niʾ...   c.3.3.08   
313  text_name    The rulers of Lagaš
Name: c.2.1.2...    c.2.1.2   
392  text_name    The history of the Tummal
Name: c...    c.2.1.3   
66   text_name    The Sumerian king list
Name: c.2....    c.2.1.1   

     lit_uniq_words  matches  
95               97       76  
78               60       47  
27               69       54  
370              99       77  
327             486      378  
250             164      127  
133              57       44  
319             665      512  
261              76       58  
214              61       46  
335             129       97  
23               51       38  
393              51       38  
129             655      488  
283             818      609  
106             116       83  
389             106       75  
320              68       48  
255              56       39  
81               60       39  
313             208      130  
392              56       29  
66              374       84

#### Observations
Generally, the literary texts with more unique words overall match the lexical corpus better than the literary works which have fewer unique words. It is interesting that no single literary composition jumps out as having comparative more words in common with the lexical corpus.

---
The literary texts with few unique words overall also have the highest percentages of matches compared to those texts. It may be wise to remove such texts from further analysis.

Many texts (114 / 319) have a 90% or greater matching score with the lexical corpus. Only 23 texts have matching less than 80%. For example, the Sumerian King List has only a 22% match (the least) but has many (374) unique words. This is probably due to the many proper names. More consideration should go into why some of these other texts have so many words that do not appear in the lexical corpus

### Compare all texts individually

Kick out lexical texts that have too few words

In [97]:
min_words = 50
lit_tl_d = {}
for lit_t in lit_tl:
    lit_wl = set(etcsl[etcsl['id_text'] == lit_t]['lemma'])
    if len(lit_wl) >= min_words:
        lit_tl_d[lit_t] = lit_wl
        #print(len(lit_wl))

In [105]:
comps = []
for lex_t in lex_tl:
    #Maybe a good idea to omit OB Ki-ulutin-bi-še
    #if lex_t == 'dcclt/Q000045':
    #    continue
        
    ##Optionally omit Ea and Izi
    #if lex_t in ['dcclt/Q000050','dcclt/Q000055']:
    #    continue
    lex_wl = set(lexical[lexical['cat_num'] == lex_t]['lemma'])
    #print(lex_t + ': ' + str(len(lex_wl)))
    for lit_t in lit_tl_d: #Make sure to use the dictionary here
        lit_wl = lit_tl_d[lit_t]
        #key = lex_t + '::' + lit_t
        lex_comp,lit_comp,match_num = compare_single(lex_wl,lit_wl)
        #lex_comps[key] = lex_comp
        #lit_comps[key] = lit_comp
        comps.append({'lit_t':lit_t,'lit_name':etcsl_cat.loc[lit_t]['text_name'],'lex_t':lex_t,'lex_name':cat_df.loc[lex_t]['designation'],
                      'lex_comp':lex_comp,'lit_comp':lit_comp,'lit_uniq_words':len(lit_wl),'lex_uniq_words':len(lex_wl),
                      'matches':match_num})
df_comps = pd.DataFrame(comps)

In [117]:
#Remove Ea and Izi df_comps[~df_comps['lex_t'].isin(['Q000050','Q000055'])]
df_comps[~df_comps['lex_t'].isin(['Q000050','Q000055'])].sort_values(by='lit_comp',ascending=False)

lex_comp          lex_name    lex_t  lex_uniq_words  lit_comp  \
708   0.055130      OB Nippur Lu  Q000047             653  0.571429   
962   0.067381      OB Nippur Lu  Q000047             653  0.564103   
658   0.075038      OB Nippur Lu  Q000047             653  0.556818   
1782  0.107895   OB Nippur Kagal  Q000048             380  0.554054   
807   0.055130      OB Nippur Lu  Q000047             653  0.545455   
1898  0.078947   OB Nippur Kagal  Q000048             380  0.545455   
929   0.045942      OB Nippur Lu  Q000047             653  0.545455   
1776  0.094737   OB Nippur Kagal  Q000048             380  0.545455   
1618  0.110526   OB Nippur Kagal  Q000048             380  0.545455   
805   0.084227      OB Nippur Lu  Q000047             653  0.544554   
1804  0.094737   OB Nippur Kagal  Q000048             380  0.537313   
1910  0.152632   OB Nippur Kagal  Q000048             380  0.537037   
1754  0.134211   OB Nippur Kagal  Q000048             380  0.536842   
960   0.079632      OB Nippur Lu  Q000047             653  0.536082   
919   0.047473      OB Nippur Lu  Q000047             653  0.534483   
657   0.073507      OB Nippur Lu  Q000047             653  0.533333   
649   0.062787      OB Nippur Lu  Q000047             653  0.532468   
1764  0.073684   OB Nippur Kagal  Q000048             380  0.528302   
1831  0.073684   OB Nippur Kagal  Q000048             380  0.528302   
748   0.056662      OB Nippur Lu  Q000047             653  0.521127   
1681  0.139474   OB Nippur Kagal  Q000048             380  0.519608   
677   0.070444      OB Nippur Lu  Q000047             653  0.516854   
945   0.052067      OB Nippur Lu  Q000047             653  0.515152   
1774  0.136842   OB Nippur Kagal  Q000048             380  0.514851   
861   0.073507      OB Nippur Lu  Q000047             653  0.510638   
730   0.076570      OB Nippur Lu  Q000047             653  0.510204   
1835  0.092105   OB Nippur Kagal  Q000048             380  0.507246   
685   0.058193      OB Nippur Lu  Q000047             653  0.506667   
669   0.059724      OB Nippur Lu  Q000047             653  0.506494   
1734  0.102632   OB Nippur Kagal  Q000048             380  0.506494   
...        ...               ...      ...             ...       ...   
3316  0.029762  Weidner God List  Q000263             168  0.022124   
3351  0.011905  Weidner God List  Q000263             168  0.021505   
3512  0.029762  Weidner God List  Q000263             168  0.021097   
3303  0.011905  Weidner God List  Q000263             168  0.020000   
3377  0.047619  Weidner God List  Q000263             168  0.019950   
3243  0.041667  Weidner God List  Q000263             168  0.019774   
3327  0.023810  Weidner God List  Q000263             168  0.019512   
3427  0.047619  Weidner God List  Q000263             168  0.019048   
3428  0.011905  Weidner God List  Q000263             168  0.019048   
3379  0.005952  Weidner God List  Q000263             168  0.018868   
3252  0.023810  Weidner God List  Q000263             168  0.018605   
3307  0.005952  Weidner God List  Q000263             168  0.018519   
3468  0.005952  Weidner God List  Q000263             168  0.018519   
3274  0.053571  Weidner God List  Q000263             168  0.018000   
3521  0.023810  Weidner God List  Q000263             168  0.017778   
3533  0.005952  Weidner God List  Q000263             168  0.017544   
3297  0.005952  Weidner God List  Q000263             168  0.016667   
3412  0.005952  Weidner God List  Q000263             168  0.016667   
3287  0.023810  Weidner God List  Q000263             168  0.016194   
3425  0.029762  Weidner God List  Q000263             168  0.015480   
3371  0.017857  Weidner God List  Q000263             168  0.015385   
3238  0.011905  Weidner God List  Q000263             168  0.014815   
3366  0.005952  Weidner God List  Q000263             168  0.013514   
3399  0.017857  Weidner God List  Q000263             168  0.013158   
3341  0.005952  Weidner God Li

In [116]:
#Remove Ki-ulutin-bi-še df_comps[df_comps['lex_t'] != 'Q000045']
df_comps[df_comps['lex_t'] != 'Q000045'].sort_values(by='lex_comp',ascending=False)

lex_comp           lex_name    lex_t  lex_uniq_words  lit_comp  \
1850  0.426316    OB Nippur Kagal  Q000048             380  0.198044   
2496  0.420000       OB Nippur Ea  Q000055             700  0.359413   
1657  0.418421    OB Nippur Kagal  Q000048             380  0.212851   
2949  0.409231  OB Lu₂-azlag₂ B-C  Q000302             325  0.178046   
2303  0.408571       OB Nippur Ea  Q000055             700  0.382865   
1880  0.397368    OB Nippur Kagal  Q000048             380  0.227068   
1724  0.381579    OB Nippur Kagal  Q000048             380  0.221374   
4241  0.372754      OB Nippur Izi  Q000050             668  0.333333   
1907  0.368421    OB Nippur Kagal  Q000048             380  0.254083   
4434  0.360778      OB Nippur Izi  Q000050             668  0.294621   
2819  0.360000   OB Nippur Ura 03  Q000001             350  0.154034   
1011  0.358333       Nippur Nigga  Q000052             360  0.172691   
1715  0.357895    OB Nippur Kagal  Q000048             380  0.229730   
1733  0.357895    OB Nippur Kagal  Q000048             380  0.298246   
3142  0.356923  OB Lu₂-azlag₂ B-C  Q000302             325  0.141809   
1659  0.355263    OB Nippur Kagal  Q000048             380  0.270000   
1725  0.350000    OB Nippur Kagal  Q000048             380  0.280591   
1828  0.347368    OB Nippur Kagal  Q000048             380  0.250474   
1900  0.344737    OB Nippur Kagal  Q000048             380  0.261477   
2684  0.342857   OB Nippur Ura 03  Q000001             350  0.202703   
1863  0.342105    OB Nippur Kagal  Q000048             380  0.369318   
1706  0.342105    OB Nippur Kagal  Q000048             380  0.262097   
2526  0.341429       OB Nippur Ea  Q000055             700  0.359398   
2361  0.340000       OB Nippur Ea  Q000055             700  0.402027   
3091  0.338462  OB Lu₂-azlag₂ B-C  Q000302             325  0.224490   
3172  0.338462  OB Lu₂-azlag₂ B-C  Q000302             325  0.165414   
2939  0.338462  OB Lu₂-azlag₂ B-C  Q000302             325  0.232558   
2626  0.334286   OB Nippur Ura 03  Q000001             350  0.156627   
1254  0.333333       Nippur Nigga  Q000052             360  0.239521   
3192  0.332308  OB Lu₂-azlag₂ B-C  Q000302             325  0.215569   
...        ...                ...      ...             ...       ...   
1578  0.012097   OB Nippur Ura 04  Q000041             496  0.120000   
1330  0.012097   OB Nippur Ura 04  Q000041             496  0.103448   
3303  0.011905   Weidner God List  Q000263             168  0.020000   
3387  0.011905   Weidner God List  Q000263             168  0.039216   
3497  0.011905   Weidner God List  Q000263             168  0.029851   
3278  0.011905   Weidner God List  Q000263             168  0.036364   
3410  0.011905   Weidner God List  Q000263             168  0.024096   
3266  0.011905   Weidner God List  Q000263             168  0.024691   
3428  0.011905   Weidner God List  Q000263             168  0.019048   
3293  0.011905   Weidner God List  Q000263             168  0.031746   
3400  0.011905   Weidner God List  Q000263             168  0.024691   
3238  0.011905   Weidner God List  Q000263             168  0.014815   
3434  0.011905   Weidner God List  Q000263             168  0.032787   
3351  0.011905   Weidner God List  Q000263             168  0.021505   
2120  0.010667     OB Nippur Diri  Q000057             375  0.066667   
2033  0.010667     OB Nippur Diri  Q000057             375  0.048780   
1613  0.010081   OB Nippur Ura 04  Q000041             496  0.089286   
3379  0.005952   Weidner God List  Q000263             168  0.018868   
3297  0.005952   Weidner God List  Q000263             168  0.016667   
3468  0.005952   Weidner God List  Q000263             168  0.018519   
3366  0.005952   Weidner God List  Q000263             168  0.013514   
3365  0.005952   Weidner God List  Q000263             168  0.008621   
3494  0.005952   Weidner God List  Q000263             168  0.010204   
3412  0.005952   Weidner God List  Q000263    

#### Observations

It may be wise to look at literary texts individually...

---

It may be wise to remove Ki-ulutin-bi-še from this analysis

### Singular Views

Let us look at Lugal-e (a literary text)

In [119]:
df_comps[df_comps['lit_t'] == 'c.1.6.2'].sort_values(by='matches',ascending=False)

lex_comp                   lex_name    lex_t  lex_uniq_words  lit_comp  \
2303  0.408571               OB Nippur Ea  Q000055             700  0.382865   
4241  0.372754              OB Nippur Izi  Q000050             668  0.333333   
688   0.283308               OB Nippur Lu  Q000047             653  0.247657   
1657  0.418421            OB Nippur Kagal  Q000048             380  0.212851   
42    0.268603           OB Nippur Ura 01  Q000039             551  0.198126   
2949  0.409231          OB Lu₂-azlag₂ B-C  Q000302             325  0.178046   
1011  0.358333               Nippur Nigga  Q000052             360  0.172691   
3918  0.257796           OB Nippur Ura 02  Q000040             481  0.165997   
2626  0.334286           OB Nippur Ura 03  Q000001             350  0.156627   
1334  0.233871           OB Nippur Ura 04  Q000041             496  0.155288   
1980  0.221333             OB Nippur Diri  Q000057             375  0.111111   
365   0.219780           OB Nippur Ura 05  Q000042             364  0.107095   
3595  0.619048  OB Nippur Ki-ulutin-bi-še  Q000045              84  0.069612   
3272  0.119048           Weidner God List  Q000263             168  0.026774   

                                          lit_name    lit_t  lit_uniq_words  \
2303  Ninurta's exploits: a šir-sud (?) to Ninurta  c.1.6.2             747   
4241  Ninurta's exploits: a šir-sud (?) to Ninurta  c.1.6.2             747   
688   Ninurta's exploits: a šir-sud (?) to Ninurta  c.1.6.2             747   
1657  Ninurta's exploits: a šir-sud (?) to Ninurta  c.1.6.2             747   
42    Ninurta's exploits: a šir-sud (?) to Ninurta  c.1.6.2             747   
2949  Ninurta's exploits: a šir-sud (?) to Ninurta  c.1.6.2             747   
1011  Ninurta's exploits: a šir-sud (?) to Ninurta  c.1.6.2             747   
3918  Ninurta's exploits: a šir-sud (?) to Ninurta  c.1.6.2             747   
2626  Ninurta's exploits: a šir-sud (?) to Ninurta  c.1.6.2             747   
1334  Ninurta's exploits: a šir-sud (?) to Ninurta  c.1.6.2             747   
1980  Ninurta's exploits: a šir-sud (?) to Ninurta  c.1.6.2             747   
365   Ninurta's exploits: a šir-sud (?) to Ninurta  c.1.6.2             747   
3595  Ninurta's exploits: a šir-sud (?) to Ninurta  c.1.6.2             747   
3272  Ninurta's exploits: a šir-sud (?) to Ninurta  c.1.6.2             747   

      matches  
2303      286  
4241      249  
688       185  
1657      159  
42        148  
2949      133  
1011      129  
3918      124  
2626      117  
1334      116  
1980       83  
365        80  
3595       52  
3272       20

Let us look at OB Nippur Ura 4 (a lexical text)

In [120]:
df_comps[df_comps['lex_t'] == 'Q000041'].sort_values(by='matches',ascending=False)

lex_comp          lex_name    lex_t  lex_uniq_words  lit_comp  \
1334  0.233871  OB Nippur Ura 04  Q000041             496  0.155288   
1527  0.187500  OB Nippur Ura 04  Q000041             496  0.113692   
1584  0.179435  OB Nippur Ura 04  Q000041             496  0.161525   
1336  0.177419  OB Nippur Ura 04  Q000041             496  0.176000   
1392  0.159274  OB Nippur Ura 04  Q000041             496  0.133446   
1489  0.159274  OB Nippur Ura 04  Q000041             496  0.188095   
1505  0.157258  OB Nippur Ura 04  Q000041             496  0.148008   
1371  0.153226  OB Nippur Ura 04  Q000041             496  0.170787   
1401  0.149194  OB Nippur Ura 04  Q000041             496  0.112977   
1402  0.149194  OB Nippur Ura 04  Q000041             496  0.156118   
1588  0.143145  OB Nippur Ura 04  Q000041             496  0.306034   
1577  0.141129  OB Nippur Ura 04  Q000041             496  0.139721   
1607  0.137097  OB Nippur Ura 04  Q000041             496  0.155606   
1557  0.135081  OB Nippur Ura 04  Q000041             496  0.100752   
1383  0.135081  OB Nippur Ura 04  Q000041             496  0.135081   
1324  0.131048  OB Nippur Ura 04  Q000041             496  0.137421   
1476  0.129032  OB Nippur Ura 04  Q000041             496  0.130612   
1410  0.129032  OB Nippur Ura 04  Q000041             496  0.140351   
1439  0.120968  OB Nippur Ura 04  Q000041             496  0.149626   
1298  0.120968  OB Nippur Ura 04  Q000041             496  0.131868   
1305  0.118952  OB Nippur Ura 04  Q000041             496  0.166667   
1545  0.116935  OB Nippur Ura 04  Q000041             496  0.154667   
1537  0.114919  OB Nippur Ura 04  Q000041             496  0.188742   
1564  0.114919  OB Nippur Ura 04  Q000041             496  0.117284   
1497  0.112903  OB Nippur Ura 04  Q000041             496  0.148148   
1542  0.112903  OB Nippur Ura 04  Q000041             496  0.172308   
1440  0.110887  OB Nippur Ura 04  Q000041             496  0.183946   
1540  0.110887  OB Nippur Ura 04  Q000041             496  0.156250   
1511  0.110887  OB Nippur Ura 04  Q000041             496  0.143979   
1457  0.110887  OB Nippur Ura 04  Q000041             496  0.165165   
...        ...               ...      ...             ...       ...   
1369  0.022177  OB Nippur Ura 04  Q000041             496  0.203704   
1509  0.022177  OB Nippur Ura 04  Q000041             496  0.144737   
1405  0.022177  OB Nippur Ura 04  Q000041             496  0.192982   
1589  0.022177  OB Nippur Ura 04  Q000041             496  0.180328   
1347  0.022177  OB Nippur Ura 04  Q000041             496  0.029412   
1449  0.022177  OB Nippur Ura 04  Q000041             496  0.215686   
1441  0.022177  OB Nippur Ura 04  Q000041             496  0.207547   
1541  0.020161  OB Nippur Ura 04  Q000041             496  0.181818   
1504  0.020161  OB Nippur Ura 04  Q000041             496  0.178571   
1428  0.020161  OB Nippur Ura 04  Q000041             496  0.135135   
1614  0.020161  OB Nippur Ura 04  Q000041             496  0.196078   
1595  0.020161  OB Nippur Ura 04  Q000041             496  0.175439   
1435  0.020161  OB Nippur Ura 04  Q000041             496  0.172414   
1468  0.018145  OB Nippur Ura 04  Q000041             496  0.147541   
1420  0.018145  OB Nippur Ura 04  Q000041             496  0.160714   
1309  0.018145  OB Nippur Ura 04  Q000041             496  0.176471   
1317  0.018145  OB Nippur Ura 04  Q000041             496  0.152542   
1508  0.018145  OB Nippur Ura 04  Q000041             496  0.169811   
1600  0.018145  OB Nippur Ura 04  Q000041             496  0.160714   
1355  0.018145  OB Nippur Ura 04  Q000041             496  0.142857   
1558  0.016129  OB Nippur Ura 04  Q000041             496  0.117647   
1360  0.016129  OB Nippur Ura 04  Q000041             496  0.160000   
1586  0.016129  OB Nippur Ura 04  Q000041             496  0.142857   
1329  0.016129  OB Nippur Ura 04  Q000041             496  0.135593   
1579  0.016129  OB Nippur Ura 

### Further tasks

* consider overall word frequency
* consider multiword expressions
* consider chopping up some of these lexical lists (like Ura) into sublists
* Include OB Royal Inscriptions